revocation/crl/fetcher.go

```diff
@@ -0,0 +1,148 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package crl provides Fetcher and Cache interface and implementations for
+// fetching CRLs.
+package crl
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// MaxCRLSize is the maximum size of CRL in bytes
+//
+// The 32 MiB limit is based on investigation that even the largest CRLs
+// are less than 16 MiB. The limit is set to 32 MiB to prevent
+const MaxCRLSize = 32 * 1024 * 1024 // 32 MiB
+
+// Fetcher is an interface that specifies methods used for fetching CRL
+// from the given URL
+type Fetcher interface {
+	// Fetch retrieves the CRL from the given URL.
+	Fetch(ctx context.Context, url string) (base, delta *x509.RevocationList, err error)
+}
+
+// HTTPFetcher is a Fetcher implementation that fetches CRL from the given URL
+type HTTPFetcher struct {
+	// Cache stores fetched CRLs and reuses them until the CRL expires.
+	// If Cache is nil, no cache is used.
+	Cache Cache
+
+	httpClient *http.Client
+}
+
+// NewHTTPFetcher creates a new HTTPFetcher with the given HTTP client
+func NewHTTPFetcher(httpClient *http.Client) *HTTPFetcher {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	return &HTTPFetcher{
+		httpClient: httpClient,
+	}
+}
+
+// Fetch retrieves the CRL from the given URL
+//
+// It try to get the CRL from the cache first, if the cache is not nil or have
+// an error (e.g. cache miss), it will download the CRL from the URL, then
+// store it to the cache if the cache is not nil.
+func (f *HTTPFetcher) Fetch(ctx context.Context, url string) (base, delta *x509.RevocationList, err error) {
+	if url == "" {
+		return nil, nil, errors.New("CRL URL is empty")
+	}
+
+	if f.Cache == nil {
+		// no cache, download directly
+		return f.download(ctx, url)
+	}
+
+	// try to get from cache
+	bundle, err := f.Cache.Get(ctx, url)
+	if err != nil {
+		return f.download(ctx, url)
+	}
+
+	// check expiry
+	nextUpdate := bundle.BaseCRL.NextUpdate
+	if !nextUpdate.IsZero() && time.Now().After(nextUpdate) {
+		return f.download(ctx, url)
+	}
+
+	return bundle.BaseCRL, nil, nil
```

Comment:
The issue still exists, when cache hit, only BaseCRL is returned, when cache miss, DeltaCRL is also returned. Therefore, the crl revocation response may not be idempotent regarding the same request.

Reply:
Please have a look again. We return `*Bundle` now. Delta CRL will always be `nil` for this version.